presto-hive/src/main/java/com/facebook/presto/hive/statistics/QuickStatsProvider.java

```diff
@@ -0,0 +1,308 @@
+                    // This should not happen because we checked that the future was completed successfully
+                    log.error(e, "Failed to get value for a quick stats future which was completed successfully");
+                    throw new RuntimeException(e);
+                }
+            }
+            else {
+                // If a future did not finish, or finished exceptionally, we do not add it to the results
+                // A new query for the same partition could trigger a successful quick stats fetch for this partition
+                result.put(partitionId, empty());
+            }
+        }
+
+        return result.build();
+    }
+
+    public PartitionStatistics getQuickStats(ConnectorSession session, SemiTransactionalHiveMetastore metastore, SchemaTableName table,
+            MetastoreContext metastoreContext, String partitionId)
+    {
+        if (!isQuickStatsEnabled(session)) {
+            return empty();
+        }
+        requestCount.incrementAndGet(); // New request was made to resolve quick stats for a partition
+
+        // Check if we already have stats cached in partitionIdToQuickStatsCache. If so return from cache
+        String partitionKey = String.join("/", table.toSchemaTablePrefix().toString(), partitionId);
+
+        PartitionStatistics cachedValue = partitionToStatsCache.getIfPresent(partitionKey);
+        if (cachedValue != null) {
+            succesfulResolveFromCacheCount.incrementAndGet();
+            return cachedValue;
+        }
+
+        // Check if we already have a quick stats build in progress for this partition key
+        if (inProgressBuilds.containsKey(partitionKey)) {
+            return empty();
+        }
+
+        // If not, atomically initiate a call to build quick stats in a background thread
+        AtomicReference<CompletableFuture<PartitionStatistics>> partitionStatisticsCompletableFuture = new AtomicReference<>();
+        inProgressBuilds.computeIfAbsent(partitionKey, (key) -> {
+            partitionStatisticsCompletableFuture.set(
+                    supplyAsync(() -> buildQuickStats(partitionKey, partitionId, session, metastore, table, metastoreContext), backgroundFetchExecutor));
+
+            // Also add a hook to reap this in-progress thread if it doesn't finish in reaperExpiry seconds
+            inProgressReaperExecutor.schedule(() -> {
+                inProgressBuilds.remove(partitionKey);
+                partitionStatisticsCompletableFuture.get().cancel(true);
+            }, reaperExpiryMillis, MILLISECONDS);
+            return Instant.now();
+        });
+
+        CompletableFuture<PartitionStatistics> future = partitionStatisticsCompletableFuture.get();
+        // If a background call to build quick stats was started, wait for it to complete
+        // This way, the first query that initiated the quick stats call for this partition will wait for the stats to be built (unless it times out)
```

Comment:
Curious if there was any other discussion about this behavior? Why not let all queries looking for this data wait for stats? Or just let all queries with in-progress jobs just return empty stats? Is there a particular difference between the query that initiated vs later ones that makes this behavior is desirable?

I feel like preventing some queries from completing, but not others would look odd from the users perspective and/or more difficult to debug. If you let all queries through regardless of in-progress status, you might not benefit from the stats immediately, but at least you're not waiting the whole time for a large table, especially if the query would be quick otherwise (e.g. scanning just a tiny partition?)

OTOH, blocking all queries due to in-progress stats might lead to undesirable performance behavior, but it might be more understandable for the user that all queries are stuck in planning. It may be easier to draw the conclusion that stats quickstats collection is the culprit.

I don't have a strong opinion of which approach to take. Just wanted some discussion.

Reply:
I understand wanting to avoid blocking for the batch use case. However, for querying data without stats loaded, will having the stats loaded actually decrease E2E query times? It's likely to be query dependent - so I'm sure for some queries it might. I'm concerned then in the common case whether it's worth blocking at all.

> A user could also use this mechanism for pre-building quick stats for newly online partitions; simply do a SHOW STATS FOR (<partition query>) before bringing the partition online, and this way all new queries will benefit from the pre-cached stats

This feels like something that should be achieved via an explicit procedure call IMO, but can understand how just dumping the query into `SHOW STATS` can be easier to load the desired partition(s).

Reply:
>for online querying data without stats loaded, will having the stats loaded actually decrease E2E query times

The latency impact of not having stats resulting in a poor plan will dominate the E2E query times, so yes having quick stats loaded will help reduce this.
If this user is concerned about having their query wait for quick stats building, they can always `quick-stats.build-timeout` set to a low enough value acceptable to them (`1ms` e.g)

Reply:
I know that stats are helpful for queries for queries which execute on an especially large dataset, but if many queries run in seconds to ms without stats, then they won't really be beneficial. At the end of the day it really depends on the workload and datasets. I personally don't have a grasp on all the workloads that people are running with Presto. What I do know is that Presto's design is catered towards fast-running analytical queries over long-running ones where stats might be more critical, hence you want the overhead of stat collection to be small.

As long as it's configurable on the session side as to how long to wait for stats, I suppose the current solution is okay

Reply:
The argument of quick vs long running queries having low/high impact from stats is bit myopic since its focused on a single query. Consider the beneficial impact of having stats available to subsequent queries and lowered cluster utilization as a result. 
To allay your concerns, I will add a session control for the build-timeout (right now it is a startup property), including setting it to `0`. This way users can choose per query the behavior they want.

Reply:
> I will add a session control property for the build-timeout (right now it is a startup property), including setting it to `0`. This way users can choose per query the behavior they want.

Please document the new session property, when you can.